woocommerce/wc-admin
# Fix the react state update error on homescreen

Fixes the React Warning on dev - `Cannon update component while rendering a different component` (see screenshot)

<img width="1207" alt="Screen Shot 2021-02-05 at 5 21 51 PM" src="https://user-images.githubusercontent.com/2240960/107235078-b0918780-69fa-11eb-889f-7d9a0564912c.png">

This was mostly happening because of creating new object references in the `useSelect` causing for many component updates. Which allowed for a high probability for a state to be updated at the same time a render took place.

This was mostly fixable by a small refactor, I did add the `rememo` library to the `data` package, this allows us to memoize selectors, especially handy for the query selector that maps the id's to the objects.
`wordpress/core-data` does the same thing, see [here](https://github.com/WordPress/gutenberg/blob/master/packages/core-data/src/selectors.js#L99-L106) that's where I got the library from.

### Detailed test instructions:

- Create a new Woocommerce store and finish the onboarding wizard
- Load the home screen in develop, there should be no red warnings in the console (like the screen shot above)
- Click on `Set up payments` task, no new warnings should pop up (I did have one randomly show up, but couldn't reproduce it after).
- Do a quick smoke test of the home screen in general, make sure the notes render correctly, and the unread indicator shows up (when needed), can dismiss and revert (card correctly disappears), store alerts
